### ISSUE TITLE
Нужен ли аддон меню?

### ISSUE THREAD

**Issue by Germenzi** [H360ru/araxxes#49]
Менюшный аддон, на мой взгляд, лишний. Он служит просто посредником и сущностью для удобного перемещения интерфейса твинами (ну, насколько я понял). Также для объединения различных сцен в цельные структуры. Но это все только создает лишние зависимости и непрозрачность структуры UI, она вся получается запутанная, т.к. все перемещения, создание экземпляров, обработка нажатий, происходят через код. Это тупо неудобно связывать с игрой.

Почему я считаю аддон лишним?
- Много зависимостей
- Плохая читабельность
- Сложность построения интерфейса

Аддон, может быть и удобный, но сложный и ненужный. Что предлагаю я: https://github.com/H360ru/araxxes/commit/1acf6efba6fd01db538ff8fb2ad79d03148d7106

Чем этот способ лучше на мой взгляд?
- Одни и те же менюшки можно использовать в разных местах
- Сцены меню не заботятся о своем использовании (какой то там из принципов ооп)
- Избавление от глобального твининга

Мой пример реализован нереально грязно, это понятно. Он просто для демонстрации моей идеи.

**Comment by alex7850** [comment]
О, гуд пример. Считаю, что вполне можно сделать и сигналами из инспектора, да так будет лучше. Проблема в поведении родительских контрол нод, которые перенимают на себя относительные координаты, и бокс контейнера, так что возникают проблемы с адаптивностью. (это важно, потом на скринах продемонстрирую проблемы). Щас на днях пороюсь с твоим примером, сам надеюсь, что не прав.

Про читаемость методов надо сравнить щас каждый пункт с примерами из кода конечно. Потому что кроме инспектора отличий-то мало.
Моя главная аргументация в экземплировании меню. Рано или поздно все равно понадобятся именно "контейнерные ноды" для сцен, чтобы асинхронно (может даже многопоточно) подгружать крупные меню.

Побочное преимущество в возможности подгружать разные ориентации экранов (там где адаптивные контейнеры еще в годо 3+ хромают) Хотя это вопрос конкретной реализации и степени универсальности.

Так же, не совсем понял про:
- "Одни и те же менюшки можно использовать в разных местах."
А в моем варианте не так?
- "Избавление от глобального твининга"
Это как бы вообще уже не вопрос, глобального твининга и не будет.

**Comment by alex7850** [comment]
Да, тут скорее к контейнерам пришел из-за кривости самого годо:
- Если ты меняешь как-то меняешь дочернее меню1 или 2 через его сцену, то у тебя в сцене мейн меню оно внезапно пропорционально изменению скейлится. (т.е. надо удалять этот экземпляр менюшки и загружать в мейн заново)
Т.е. воркфлоу уже поломан для дизайнера полностью. Так бы сам рад. Мой подход с рантайм подгрузкой этого косяка лишен. Ну и менеджмент памяти нам необходим.
![изображение](https://user-images.githubusercontent.com/15798187/124357078-ef033980-dc5c-11eb-8ab5-a59ab864aa66.png)

Что касается соединения менюшек непосредственно через инспектор и сигналы, то надо вводить, 100% удобнее в большинстве случаев. Хреново только, что сигналы как объекты ввели пока только в годо 4.0, тогда еще больше гибкости.

Про упрощение ооп буду дальше думать, смотреть, конечно

**Comment by Germenzi** [comment]
Ну, у тебя все равно есть некоторая привязка к внешнему дереву, вызывая методы родителя. А с моим методом вообще что угодно, когда и как угодно. 

Насчет подгрузки больших меню не понял. Неужели UI требует долгой загрузки? Или ты про сцены?

**Comment by Germenzi** [comment]
> Если ты как-то меняешь дочернее меню1 или 2 через его сцену, то у тебя в сцене мейн меню оно внезапно 
> пропорционально изменению скейлится.

Не понял. Там я у менюшек в мейн сцене поставил скейл x3, потому что там у тебя что то в настройках стоит и они очень маленькие были.

**Comment by Germenzi** [comment]
Можешь как то гифкой или скрином показать?

**Comment by alex7850** [comment]
Почему-то не могу воспроизвести, хоть перед изменениями поставил скейл на 1 1.

Кстати, есть очень стойкое предчувствие, что контролы пофиксили, вообще иначе все себя ведет. Возможно, потому что сейчас в сценах у тебя пока один контейнер и его нет в корне сцен меню. Надо еще будет посмотреть.

**Comment by Germenzi** [comment]
А что именно было не так? Разве так и не должно быть, что меняя класс, меняется его экземпляр?

**Comment by alex7850** [comment]
> Неужели UI требует долгой загрузки? Или ты про сцены?
Да, может быть серьезный лагспайк прямо, есть крайние случаи, а еще это все весит 100% времени в сцене и памяти. Т.е. в браузерном, мобильном экспорте без таких подзагрузчиков какие-то внутренние вики, туториалы, статистику юнитов не сделать. (там по несколько сотен, если не тысяч объектов и самих нод на менюшку же)

**Comment by Germenzi** [comment]
В таком случае можно поступить банально и сделать экран подзагрузки, или просто иконку по типу "загрузка" где нибудь внизу экрана.

**Comment by alex7850** [comment]
И будешь каждый раз ждать на старте, чтобы браузер прогрузил все менюшки?

**Comment by Germenzi** [comment]
Не все. Ну например нажимаешь ты на статистику, и какое то время висит иконка ожидания. По моему почти везде так делают

**Comment by alex7850** [comment]
Ну это да, это и есть как бы контейнеры, которые я сделал.

Сравнил полную загрузку в память против "по необходимости".
Результаты интересные.
По объектам разница оказалась меньше, чем по нодам, внезапно.
Объекты 1400 - 1200
Ноды 250 - 120

По памяти особой разницы нет, конечно, зато экземплирование нод в дерево достаточно долгая штука.

**Comment by Germenzi** [comment]
Ну насчет твоих контейнеров не знаю, я думаю можно сделать и проще. Главное меню или какой то синглтон в несколько потоков грузят сцену, тут никакие аддоны не нужны.

**Comment by alex7850** [comment]
С упрощением и рефакторингом еще заведомо был согласен. Особенно ответственностей класса за анимацию и подзагрузку одновременно.

Но это сейчас вопрос даже не самый важный. Реализации логики через иерархию дерева меню или плоской организации через сигналы.

В моем случае минус в необходимости все равно цеплять к сцене такой же как у тебя скрипт с реакцией на кнопки.
(Вообще планировалось сделать какой-то класс кнопок, которым в инспекторе просто задаешь ссылку на нужное меню и не нужны никакие сигналы через инспектор)
![изображение](https://user-images.githubusercontent.com/15798187/124359391-0e539400-dc68-11eb-9ca6-71f87c77e83b.png)

**Comment by Germenzi** [comment]
> Вообще планировалось сделать какой-то класс кнопок, которым в инспекторе просто задаешь ссылку на нужное меню и не > нужны никакие сигналы через инспектор

Это ты опять, как по мне, придумываешь. Подключить сигнал к функции дело 3-х секунд, а давать ссылку кнопке на меню, это опять же нечитабельно и порождает зависимости. Например, меню обязательно должно уметь работать с каждым типом кнопок, обрабатывать их запросы.

> Ну и у разных меню, кстати, разный тип перехода анимации может быть, хотелось бы такую гибкость сохранить.

Разный тип анимации это ж вроде совсем просто, если ты про типы твининга конечно. Можно сделать экспорт переменную и использовать ее во всех вызовах твина.
